### ISSUE TITLE
Node.js stream support missing in LargFileUploadTask

### ISSUE THREAD

**Issue by naugtur** [microsoftgraph/msgraph-sdk-javascript#320]
# Bug Report

## Prerequisites

-   [ x ] Can you reproduce the problem?
-   [ x ] Are you running the latest version?
-   [ x ] Are you reporting to the correct repository?
-   [ x ] Did you perform a cursory search?
        yes, I found a gist that suggested reading entire file into memory for each chunk splicing :)

## Description

Attempting to pass a Stream into LargeFileUploadTask as file content is failing with
```
this.file.content.slice is not a function
```

```js
const fileObject = {
    size,
    content: fileStream,
    name: targetFileName,
  };
  const uploadTask = await new LargeFileUploadTask(
    graphClient,
    fileObject,
    uploadSession
  );
  const response = await uploadTask.upload()
```

## Steps to Reproduce

1. Take the example https://github.com/microsoftgraph/msgraph-sdk-javascript/blob/HEAD/docs/tasks/LargeFileUploadTask.md
2. use `var readStream = fs.createReadStream(filename);` to get a file stream and pass that in as content

**Expected behavior:** [What you expected to happen]

Stream flows and gets accumulated into chunks that get uploaded

**Actual behavior:** [What actually happened]

Error stemming from the assumption content is a Buffer

## Additional Context

Streams allow for a container limited to 100MB of RM to upload a multi GB file successfully. 
LargFileUploadTask is where it breaks.

Users of the SDK can't afford to load the entire file into RAM if they're building a multi tenant web app.

I'm willing to contribute a solution if I get someone from the maintainers to help me out (I have little experience using TypeScript and typing a stream is difficult)

## Usage Information

SDK Version - [SDK version you are using]

-   [ x ] Node (Check, if using Node version of SDK)

> Node Version - [The version of Node you are using]

10+ (I used multiple)
 
[AB#7546](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/7546)

**Comment by naugtur** [comment]
Here's what I'd like to do:
1. If it's possible to send entire upload to the session in one chunk (as discussed on Twitter) I'd like to expose a method to do that, which would consume anything the http client can put into body (string, blob, buffer, formdata, but most of all Stream)
2. If the file content passed to chunk upload helper is a stream:
 - prepare two buffers (pooling instead of garbage collection 🎉)
 - consume the stream
 - collect it in a buffer as it flows
 - cut off at the chunk size and continue streaming to the other buffer
 - take the full buffer and send it as a chunk
 - apply backpressure to the source stream while waiting for the upload request to process; keep the full buffer handy for retries in case they're needed. Discard only on success
 
to implement backpressure a few tricky details in a Writable implementation would need to happen.

I'd build it in JS and then need help converting to TS
Also, testing would probably have to be end-to-end, because it's easy to get invalid results if you start mocking stream consumption instead of actually sending it over the network.

Action items:
Please let me know where to look for docs or code explaining (1)
Please let me know if you're interested in working on 2 and someone can allocate a few hours to help me out.

**Comment by MIchaelMainer** [comment]
@naugtur We'd be happy to have you contribute here. @nikithauc will work with you on this.

**Comment by naugtur** [comment]
Ok. Allocating time next week!

**Comment by MIchaelMainer** [comment]
Additional reference PRs:

https://github.com/microsoftgraph/msgraph-sdk-dotnet/pulls?q=is%3Apr+author%3Aandrueastman+large+is%3Aclosed
https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/pull/65

**Comment by naugtur** [comment]
Hi @nikithauc
Could you share info or docs about the option to send a single chunk to the upload session? 

Darrel Miller mentioned that on Twitter.
> Thanks for opening the issue.  We should definitely investigate how to enable streaming bytes into range requests.  We should also investigate if just doing a plain transfer-encoding: chunked request would also work to avoid range requests.

 I'd like to start with proving I can make it work.

**Comment by nikithauc** [comment]
@naugtur   Currently it is not possible to send a single chunk in the  upload session and the way send a stream is through range requests.

**Comment by naugtur** [comment]
I can confirm the uploadSession demands all details about ranges. 

Sending 
`Content-Range: 0-${N-1}/${N}`  
works for a single chunk upload of a file larger than 4MB though. 
I tried the request directly, not via UploadTask.

Then I also uploaded 203MB powerpoint file as one chunk and it worked fine. 
Any thoughts on that?
The only limitation on chunk size was a popular JS library that's a dependency in many http clients, but I switched it off and the API itself gobbled up 200MB and crumbs.

@nikithauc Can you ask someone internally and see if they're surprised I uploaded 200MB?

**Comment by MIchaelMainer** [comment]
> uploaded 203MB... as one chunk and it worked fine. 

I'm surprised as the max per [documentation is 60MB](https://docs.microsoft.com/en-us/graph/api/driveitem-createuploadsession?view=graph-rest-beta#upload-bytes-to-the-upload-session).

> You can upload the entire file, or split the file into multiple byte ranges, as long as the maximum bytes in any given request is less than 60 MiB.

@ificator do we need to update the documentation on this one?

**Comment by naugtur** [comment]
I'm happy with how it works but if it is unintended, please let me know and I'll avoid abusing it :)

I uploaded the file into a SharePoint site, if that changes anything.

Planning to work on converting stream to chunks anyway.

**Comment by baywet** [comment]
I've been working on this on the Java SDK recently. The upload services behave a bit differently between OneDrive consumer (live account) OneDrive for business, SharePoint and file attachments on Outlook. It worth testing all these scenarios to catch eventual issues.

**Comment by nikithauc** [comment]
@MIchaelMainer -> Came across this [Stack Overflow post](https://stackoverflow.com/questions/3571115/sharepoint-2010-iis-7-5-byte-range-request-responds-with-entire-file). The post mentions a download request but could it be the related  to why a 203 MB chunk got uploaded as a single request?

**Comment by nikithauc** [comment]
@naugtur  I know that the discussion for this issue is in the initial stage but as we move to the implementation design phase, I would like to work with you on the design part of this issue so that we have it in sync with the rest of MS graph SDKs . 

Thanks,
Nikitha

**Comment by naugtur** [comment]
@nikithauc obviously!
I got in touch because I don't want to slap up a PR that never gets merged, I'd rather do the node streams specific part and allow someone else to put it in. Anyway, I'm prepared to fulfill this expectation.

As part of the process I mean to implement the good parts with streams first and show you.
As a starting point.

Meanwhile my team is on a deadline and I need to solve it for them so that an app with limited RAM can upload a 200 MB pptx to SharePoint for a few hundred users. So I went and tried what's the least I can do to make it happen too ;)

**Comment by nikithauc** [comment]
`You can upload the entire file, or split the file into multiple byte ranges, as long as the maximum bytes in any given request is less than 60 MiB.` Considering this line from the [OneDrive upload documentation](https://docs.microsoft.com/en-us/graph/api/driveitem-createuploadsession?view=graph-rest-beta#upload-bytes-to-the-upload-session) , what I would do is allow the user to decide what best works for their app/requirements.  We can allow the user to upload using range requests or upload in a single chunk. The size of the range for the requests can be decided by the user in the LargeFileUploadTask src file. I know [OneDriveLargeFileUploadTask](https://github.com/microsoftgraph/msgraph-sdk-javascript/blob/aaf142eff2c6a73af8624e0530c27e0bc04f3e59/src/tasks/OneDriveLargeFileUploadTaskUtil.ts#L40) specifies 60MB range limit and I would not change that because the its better to go with the documentation now. 

@MIchaelMainer would like to know your thoughts on this?

**Comment by MIchaelMainer** [comment]
@nikithauc I'm in agreement. Our best bet is to use the documentation as spec as much as possible. With that said, it is a challenge to provide prescriptive guidance as limits can change without us knowing. I'd expect an implementer of large file upload to test different range sizes to determine what is optimal for their scenario. So, we should leave it to the implementer to specify their range size based on their testing. I can imagine that the optimal range size will be different for an application that occasionally uploads a file for a single account versus an application that uploads thousands of files across many accounts and across many organizations. Even different geo-locations may be serviced by different clouds that have a different number of servers so that may be a factor. Applications that are uploading many large files may want to consider implementing heuristics to adjust the range size for subsequent uploads based on the performance on past uploads. I've gone on a bit of a tangent... 

We also need to make sure that we support large file attachment upload and the potential for future service APIs to support a similar mechanism.

**Comment by nikithauc** [comment]
@naugtur. Please let me know if you need more information for this issue.

**Comment by naugtur** [comment]
I had a lot on my plate at work recently and simply didn't get down to this.

The only information I could use additionally is which node versions you intend to support. There was a shift in stream best practices fairly recently and I need to know which if the new stuff I can use.

**Comment by nikithauc** [comment]
@naugtur  Sorry for a super delay in the response! 
Currently, we are not specific on the version support, but the idea would be to have something that would be flexibly working in multiple node versions, 
I am aware that SharePoint library does not support some versions and they use Graph JS SDK. 
https://docs.microsoft.com/en-us/sharepoint/dev/spfx/set-up-your-development-environment

**Comment by fabiofenoglio** [comment]
I am really interested in this too. In case someone needs it, I am using this drop-in alternative to the default LargeFileUploadTask in order to allow an upload from a generic Stream:
https://gist.github.com/fabiofenoglio/388a80204c2f203893882644567205f7

usage:
```
const yourBigFile = '/your-filesystem/big-file.mp4';
const pathToUploadIn = 'Test';
const fileNameToUpload = 'huge-file.mp4';

// find folder to upload in
const uploadFolder = await client.api(`/me/drive/root:/${pathToUploadIn}`).get() as MicrosoftGraph.DriveItem;

// create the upload session
const createUploadSessionUrl = `/me/drive/items/${uploadFolder.id}:/${encodeURIComponent(fileNameToUpload)}:/createUploadSession`;
const largeFileUploadSession = await LargeFileUploadTask.createUploadSession(
  client,
  createUploadSessionUrl,
  {
    item: {
      '@microsoft.graph.conflictBehavior': 'fail',
      name: fileNameToUpload,
    },
  },
);

// read file info and prepare the upload object.
// 'size' could be retrieved in any other way.
// contentProvider could not only read from filesystem but from any stream
const fileInfo = fs.statSync(yourBigFile);

const fileObject = {
  size: fileInfo.size,
  contentStream: fs.createReadStream(yourBigFile),
  name: fileNameToUpload,
};

// prepare the upload task.
// includes a callback for each slice upload if the uploadingSliceCallback option is given.
const uploadTask = new LargeFileUploadStreamTask(
  client,
  fileObject,
  largeFileUploadSession, {
    uploadingSliceCallback: (range) => {
      const percentage = 100.0 * range.minValue / fileInfo.size;
      console.log(`uploading: ${Math.floor(percentage)}%`)
    }
  }
);

// launch the upload and wait for completion
const response = await uploadTask.upload();
```

**Comment by naugtur** [comment]
I got busy building user value. Is inactivity the reason this is now closed?
There's still a plan to make this happen. If you're interested in getting that code, we want to share. If not, we can keep it private.

**Comment by nikithauc** [comment]
@naugtur sorry I missed you last comment. 
This issue was closed as the PR was merged and is awaiting to be released. 

The  support for  Node.js stream large file upload will be released in 3.0.0 version of the JS SDK.

@naugtur @fabiofenoglio I would appreciate if you can try the  [3.0.0- Preview.1](https://www.npmjs.com/package/@microsoft/microsoft-graph-client/v/3.0.0-Preview.1) and provide a feedback on the new  `StreamUpload` class.

```
const fileName = "<FILE_NAME>";
const stats = fs.statSync(`./test/sample_files/${fileName}`);
const totalsize = stats.size;
const readStream = fs.createReadStream(`./test/sample_files/${fileName}`);
const fileObject = new StreamUpload(readStream, fileName, totalsize);

const uploadTask = new LargeFileUploadTask(client, fileObject, uploadSession, optionsWithProgress);
const uploadResult: UploadResult = await uploadTask.upload();
``` 
Learn more about this from the updated [LargeFileUpload documentation](https://github.com/microsoftgraph/msgraph-sdk-javascript/blob/dev/docs/tasks/LargeFileUploadTask.md).

**Comment by fabiofenoglio** [comment]
@nikithauc thank you very much for your feedback. I will test this as soon as possible but it will take a few days before I'll be able to get back on this.
have a nice day!
